### DISCUSSION TITLE
base-adc-a2dp ---> basic-a2dp-audioi2s ... no data receive at receiver

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#935] [Q&A] [answered]
use above 2 fle wth below changes 
1:       //#include "AudioConfigLocal.h"
2:       make the naming same for A2DP sender and receiver  . 

no data receive at receiver . seem esps are not connecting . 

pre test : receiver is connecting and receive data from mobile .  ( while perform above test of two esps  i keep the mobile switch off so that esp not connect with mobile )  so may be some thing with sender . 

attaches is the exact sketch  I used .
[sender.txt](https://github.com/pschatzmann/arduino-audio-tools/files/12102398/sender.txt)
[receaver.txt](https://github.com/pschatzmann/arduino-audio-tools/files/12102400/receaver.txt)

**Comment by aqildad-create** [comment]
after changing the sender/source  to below then its working and receiving data ? so what's wrong with base sender/source ? 


#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

const uint16_t sample_rate = 44100;
AnalogAudioStream in; 
 
// A2DPStream a2dpStream = A2DPStream::instance(); // access A2DP as stream
A2DPStream a2dpStream;    
VolumeStream volume(a2dpStream);
StreamCopy copier(volume, in); // copy i2sStream to a2dpStream
ConverterAutoCenter<int16_t> converter;

void setup() {
  // put your setup code here, to run once:
Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
      volume.setVolume(22.3);
    
    // start bluetooth
    Serial.println("starting A2DP...");
    auto cfgA2DP = a2dpStream.defaultConfig(TX_MODE);
    cfgA2DP.name = "a2dp-i2s";
    a2dpStream.begin(cfgA2DP);


  auto cfgRx = in.defaultConfig(RX_MODE);
  cfgRx.sample_rate = sample_rate;
  in.begin(cfgRx);
 }

void loop() {
  // put your main code here, to run repeatedly:
copier.copy(converter);
}

**Comment by aqildad-create** [comment]
above base-a2dp still not connecting .  

I  have INMP441  and below code s connecting success  with above base receiver / sink  but sound very distorted . 

I feel  their is some wrong with my i2s configuration or conversion   .

below is the code  of sender 



/**
 * @file streams-i2s-a2dp.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-a2dp/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

I2SStream i2sStream;                            // Access I2S as stream
A2DPStream a2dpStream;                          // access A2DP as stream
VolumeStream volume(a2dpStream);
//StreamCopy copier(volume, conv); // copy i2sStream to a2dpStream
ConverterFillLeftAndRight<int16_t> filler(LeftIsEmpty); // fill both channels
FormatConverterStream conv(a2dpStream);
//FormatConverterStream conv(a2dpStream);

StreamCopy copierr(conv, i2sStream);
// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);

    // set intial volume
    volume.setVolume(0.3);
    
    // start bluetooth
    Serial.println("starting A2DP...");
    auto cfgA2DP = a2dpStream.defaultConfig(TX_MODE);
    cfgA2DP.name = "MyReceiver";
    cfgA2DP.auto_reconnect = true;
    a2dpStream.begin(cfgA2DP);





   
  auto config_in = i2sStream.defaultConfig(RX_MODE);
  config_in.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT | I2S_STD_FORMAT
  config_in.is_master = true;
  config_in.use_apll = false;  // try with yes
  config_in.pin_mck = 3;
  config_in.bits_per_sample = 32; 
  config_in.channels = 2; 
  config_in.sample_rate = 44100; 
  i2sStream.begin(config_in);
  AudioInfo from(44100, 2, 32);
  AudioInfo to  (44100, 2, 16);

  conv.begin(from, to);





}

// Arduino loop - copy data
void loop() {
   // copier.copy(filler);
    copierr.copy();
//    copier.copy(filler);
}

**Comment by aqildad-create** [comment]
below s base -a2dp 
why it is not connecting also why it is continuous resetting after some time ? 
/**
 * @file base-i2s-a2dp.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/base-i2s-a2dp/README.md
  * @copyright GPLv3
*/

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

/**
 * @brief We use a INMP441 I2S microphone as input and send the data to A2DP
 * Unfortunatly the data type from the microphone (int32_t)  does not match with the required data type by A2DP (int16_t),
 * so we need to convert.
 */ 

BluetoothA2DPSource a2dp_source;
I2SStream i2s;
ConverterFillLeftAndRight<int16_t> bothChannels(LeftIsEmpty);
const size_t max_buffer_len = 150;
const int channels = 2;
const size_t max_buffer_bytes = max_buffer_len * sizeof(int16_t) * channels;
uint8_t buffer[max_buffer_bytes]={0};

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t len) {
  size_t req_bytes = min(max_buffer_bytes, len*2*sizeof(int16_t));
  // the microphone provides data in int32_t -> we read it into the buffer of int32_t data so *2
  size_t result_bytes = i2s.readBytes((uint8_t*)buffer, req_bytes*2);
  // we have data only in 1 channel but we want to fill both
  return bothChannels.convert((uint8_t*)buffer, result_bytes);
}


// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  //AudioLogger::instance().begin(Serial, AudioLogger::Info);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);
  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  cfg.is_master = true;
  i2s.begin(cfg);

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.set_auto_reconnect(false);
  a2dp_source.start("MyReceiver", get_sound_data);  
}

// Arduino loop - repeated processing 
void loop() {
  delay(1000);
  bool don;
  
}

**Comment by pschatzmann** [comment, marked as answer]
I am pretty sure you left the speaker in a state that would not accept any connections: See https://github.com/pschatzmann/ESP32-A2DP/discussions/439
